### DISCUSSION TITLE
Report fields

### DISCUSSION THREAD

**Discussion by robbagithub** [netbox-community/netbox#9446] [Q&A]
Hi, 

I'm trying to write some reports and I'm having difficulty finding somewhere that shows me the options that exist to call or enquire about the object properties. 

For example I'm looking at the device racking report and want to prevent it from reporting on PDUs, but I'm not sure how I go about referencing that property. 

I've looked at the API links but I don't see where this detail would be. 

I went poking through the model files on /lfs/global/apps/netbox/netbox/dcim/models but it seems a difficult place to list all the options. 

Any help gratefully received.

Regards,
Rob.

**Comment by AndMrzv** [comment]
With reports you can call Device.objects or DeviceRole.objects or DeviceType.objects depending on what suits your task best in your python code with filter() to limit what you want to get or exclude() for what you want to avoid.
For example if you have PDUs as a Device Role with id = 7 and want everything but them, it will be Device.objects.exclude(device_role__id=7).
There is a sample at https://docs.netbox.dev/en/stable/customization/reports/ And django docs cover much more: https://docs.djangoproject.com/en/4.0/ref/models/querysets/
Hope this helps.

**Reply by robbagithub** [reply]
Thanks very much @AndMrzv i'll take a deeper look into the djangoproject details.

**Reply by robbagithub** [reply]
```
from dcim.choices import DeviceStatusChoices
from dcim.models import Device, Rack,DeviceType
from extras.reports import Report

class DeviceRackingReport(Report):
    description = "Verify each device is assigned to a Rack"
    def test_device_racking(self):
        for device in Device.objects.filter(status=DeviceStatusChoices.STATUS_ACTIVE):
            if device.rack_id is not None:
                if Device.objects.exclude(device_role_id!=1006):
                    if device.position is not None:
                        self.log_success(device)

                    elif (device.device_type.is_child_device):
                        self.log_info(device, "Device is child device and therefore not racked itself")

                    else:
                        self.log_warning(device, "Device is racked, but not assigned a position")

                else:
                    self.log_failure(device, "Device is not racked")
            else:
                self.log("It's a PDU")
```


An exception occurred: NameError: name 'device_role_id' is not defined Traceback (most recent call last): File "/lfs/global/apps/netbox-3.2.2/netbox/extras/reports.py", line 229, in run test_method() File "/lfs/global/apps/netbox-3.2.2/netbox/reports/Dev2.py", line 10, in test_device_racking if Device.objects.exclude(device_role_id!=1006): NameError: name 'device_role_id' is not defined

I've looked through the models in the source code but find no reference to device_role_id!? 
I'm sure netbox used to have a page where you could view the models, however I think it's back to django school for me !

**Reply by kkthxbye-code** [reply]
`if Device.objects.exclude(device_role_id!=1006):`

That wont work. 

`if Device.objects.exclude(device_role_id=1006):`

If it is the device role with id 1006 you want to remove from the query.
